operator/validator/metadata/syncer.go

```diff
@@ -0,0 +1,285 @@
+		logger:            logger,
+		shareStorage:      shareStorage,
+		beaconNetwork:     beaconNetwork,
+		fetcher:           NewFetcher(logger, beaconNode),
+		syncInterval:      defaultSyncInterval,
+		streamInterval:    defaultStreamInterval,
+		updateSendTimeout: defaultUpdateSendTimeout,
+	}
+
+	for _, opt := range opts {
+		opt(u)
+	}
+
+	return u
+}
+
+type Option func(*ValidatorSyncer)
+
+func WithSyncInterval(interval time.Duration) Option {
+	return func(u *ValidatorSyncer) {
+		u.syncInterval = interval
+	}
+}
+
+func (u *ValidatorSyncer) SyncOnStartup(ctx context.Context) (map[spectypes.ValidatorPK]*beacon.ValidatorMetadata, error) {
+	// Load non-liquidated shares.
+	shares := u.shareStorage.List(nil, registrystorage.ByNotLiquidated())
+	if len(shares) == 0 {
+		u.logger.Info("could not find non-liquidated validator shares on initial metadata retrieval")
+		return nil, nil
+	}
+
+	needToSync := false
+	allPubKeys := make([]spectypes.ValidatorPK, 0, len(shares))
+	for _, share := range shares {
+		allPubKeys = append(allPubKeys, share.ValidatorPubKey)
+		if !share.HasBeaconMetadata() {
+			needToSync = true
+		}
+	}
+
+	if !needToSync {
+		// No need to fetch metadata if all shares have it. It's going to be updated by Stream method afterwards.
+		return nil, nil
+	}
+
+	// Sync all pubkeys. We don't need to batch them because we need to wait here until all of them are synced.
+	return u.Sync(ctx, allPubKeys)
+}
+
+func (u *ValidatorSyncer) Sync(ctx context.Context, pubKeys []spectypes.ValidatorPK) (Validators, error) {
+	fetchStart := time.Now()
+	metadata, err := u.fetcher.Fetch(ctx, pubKeys)
+	if err != nil {
+		return nil, fmt.Errorf("fetch metadata: %w", err)
+	}
+
+	u.logger.Debug("🆕 fetched metadata for validator shares",
+		fields.Took(time.Since(fetchStart)),
+		zap.Int("metadata_cnt", len(metadata)),
+		zap.Int("shares_cnt", len(pubKeys)),
+	)
+
+	updateStart := time.Now()
+	// TODO: Refactor share storage to support passing context.
+	if err := u.shareStorage.UpdateValidatorsMetadata(metadata); err != nil {
+		return metadata, fmt.Errorf("update metadata: %w", err)
+	}
+
+	u.logger.Debug("🆕 updated validators metadata in storage",
+		fields.Took(time.Since(updateStart)),
+		zap.Int("metadata_count", len(metadata)),
+		zap.Int("shares_cnt", len(pubKeys)),
+	)
+
+	return metadata, nil
+}
+
+type ValidatorUpdate struct {
+	IndicesBefore []phase0.ValidatorIndex
+	IndicesAfter  []phase0.ValidatorIndex
+	Validators    Validators
+}
+
+func (u *ValidatorSyncer) Stream(ctx context.Context) <-chan ValidatorUpdate {
+	metadataUpdates := make(chan ValidatorUpdate)
+
+	go func() {
+		defer close(metadataUpdates)
+
+		for {
+			update, done, err := u.prepareUpdate(ctx)
+			if err != nil {
+				u.logger.Warn("failed to prepare validators metadata",
+					zap.Error(err),
+				)
+				if slept := u.sleep(ctx, u.streamInterval); !slept {
+					return
+				}
+				continue
+			}
+
+			if len(update.Validators) == 0 {
+				continue
```

Comment:
wouldn't this be a busy loop that keeps requesting non stop when there are 0 validators? I guess we need some backoff..

Reply:
I added a sleep